src/Covenant/Internal/KindCheck.hs

```diff
@@ -71,13 +84,32 @@ lookupDeclaration tn = do
     Nothing -> throwError $ UnknownType tn
     Just decl -> pure decl
 
+{- This sanity checks datatype declarations using the criteria enumerated above.
+
+-}
+
+checkDataDecls :: Map TyName (DataDeclaration AbstractTy) -> Either KindCheckError ()
+checkDataDecls decls = runKindCheckM decls $ traverse_ checkDataDecl (M.elems decls)
+
+-- REVIEW/TODO: Maybe we should do the encoding strategy sanity check here too?
```

Comment:
I'm OK with keeping them separate. Yes it's a second traversal, but I don't think we're that worried.